src/Build/Evaluation/Expander.cs

```diff
@@ -645,68 +644,89 @@ private static string[] ExtractFunctionArguments(IElementLocation elementLocatio
 
-                                if (n == -1)
-                                {
-                                    ProjectErrorUtilities.ThrowInvalidProject(elementLocation, "InvalidFunctionPropertyExpression", expressionFunction, AssemblyResources.GetString("InvalidFunctionPropertyExpressionDetailMismatchedQuote"));
-                                }
+                        n = ScanForClosingQuote(argumentsString[quoteStart], argumentsString, n);
 
-                                argumentBuilder.Append(argumentsString, quoteStart, (n - quoteStart) + 1);
-                            }
-                            else if (argumentsContent[n] == ',')
-                            {
-                                // We have reached the end of the current argument, go ahead and add it
-                                // to our list
-                                AddArgument(arguments, argumentBuilder);
+                        if (n == -1)
+                        {
+                            ProjectErrorUtilities.ThrowInvalidProject(elementLocation, "InvalidFunctionPropertyExpression", expressionFunction, AssemblyResources.GetString("InvalidFunctionPropertyExpressionDetailMismatchedQuote"));
+                        }
 
-                                // Clear out the argument builder ready for the next argument
-                                argumentBuilder.Remove(0, argumentBuilder.Length);
-                            }
-                            else
-                            {
-                                argumentBuilder.Append(argumentsContent[n]);
-                            }
+                        if (argumentStartIndex.HasValue)
+                        {
+                            argumentBuilder.Append(argumentsString, argumentStartIndex.Value, quoteStart - argumentStartIndex.Value);
+                            argumentStartIndex = null;
                         }
+                        argumentBuilder.Append(argumentsString, quoteStart, (n - quoteStart) + 1);
                     }
+                    else if (argumentsString[n] == ',')
+                    {
+                        if (argumentStartIndex.HasValue)
+                        {
+                            argumentBuilder.Append(argumentsString, argumentStartIndex.Value, n - argumentStartIndex.Value);
+                            argumentStartIndex = null;
+                        }
+
+                        // We have reached the end of the current argument, go ahead and add it
+                        // to our list
+                        AddArgument(arguments, argumentBuilder);
+
+                        // Clear out the argument builder ready for the next argument
+                        argumentBuilder.Clear();
+                    }
+                    else
+                    {
+                        argumentStartIndex ??= n;
+                    }
+                }
+
+                if (argumentStartIndex.HasValue)
```

Comment:
The original code was appending to the `argumentBuilder` character by character. With the new data structure this would be inefficient so now it keeps track of the start of the current argument and appends the entire substring when it finds its end. This `if` statement is handling the case where the argument's last character is the last character of the input string so we don't detect the end inside the loop.